File: lib/services/api_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../config/api_config.dart';

/// Cliente HTTP para comunicar el frontend Flutter con la API Flask.
/// Usa [ApiConfig.baseUrl] para todas las peticiones.
/// Si hay token ([setAuthToken]), se envía como header Authorization: Bearer.
class ApiClient {
  ApiClient._();

  static final _client = http.Client();
  static String? _authToken;

  static void setAuthToken(String? token) {
    _authToken = token;
  }

  /// True si hay JWT guardado (p. ej. para preferir `GET .../me` frente a `/perfil?email=`).
  static bool get hasAuthToken =>
      _authToken != null && _authToken!.trim().isNotEmpty;

  static Map<String, String> get _headers {
    final h = <String, String>{
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };
    if (_authToken != null && _authToken!.isNotEmpty) {
      h['Authorization'] = 'Bearer $_authToken';
    }
    return h;
  }

  static String _url(String path) {
    final p = path.startsWith('/') ? path : '/$path';
    return '${ApiConfig.baseUrl}$p';
  }

  /// GET a [path]. Ej: get('/productos')
  static Future<ApiResponse> get(String path) async {
    try {
      final res = await _client
          .get(Uri.parse(_url(path)), headers: _headers)
          .timeout(Duration(seconds: ApiConfig.timeoutSeconds));
      return _handleResponse(res);
    } catch (e, st) {
      return ApiResponse.error(e.toString(), stackTrace: st);
    }
  }

  /// POST a [path] con body [data] (se serializa a JSON).
  static Future<ApiResponse> post(String path, [Map<String, dynamic>? data]) async {
    try {
      final body = data != null ? jsonEncode(data) : null;
      final res = await _client
          .post(Uri.parse(_url(path)), headers: _headers, body: body)
          .timeout(Duration(seconds: ApiConfig.timeoutSeconds));
      return _handleResponse(res);
    } catch (e, st) {
      return ApiResponse.error(e.toString(), stackTrace: st);
    }
  }

  /// PUT a [path] con body [data].
  static Future<ApiResponse> put(String path, [Map<String, dynamic>? data]) async {
    try {
      final body = data != null ? jsonEncode(data) : null;
      final res = await _client
          .put(Uri.parse(_url(path)), headers: _headers, body: body)
          .timeout(Duration(seconds: ApiConfig.timeoutSeconds));
      return _handleResponse(res);
    } catch (e, st) {
      return ApiResponse.error(e.toString(), stackTrace: st);
    }
  }

  /// DELETE a [path].
  static Future<ApiResponse> delete(String path) async {
    try {
      final res = await _client
          .delete(Uri.parse(_url(path)), headers: _headers)
          .timeout(Duration(seconds: ApiConfig.timeoutSeconds));
      return _handleResponse(res);
    } catch (e, st) {
      return ApiResponse.error(e.toString(), stackTrace: st);
    }
  }

  static ApiResponse _handleResponse(http.Response res) {
    final ok = res.statusCode >= 200 && res.statusCode < 300;
    dynamic body;
    try {
      body = res.body.isEmpty ? null : jsonDecode(res.body);
    } catch (_) {
      body = res.body;
    }
    if (ok) {
      return ApiResponse.success(res.statusCode, body);
    }
    final String msg = body is Map
        ? (body['message'] ?? body['error'] ?? body['msg'] ?? res.body).toString()
        : res.body;
    return ApiResponse.error(msg, statusCode: res.statusCode, data: body is Map ? body : null);
  }
}

/// Respuesta unificada de la API.
class ApiResponse {
  ApiResponse._({
    this.ok = false,
    this.statusCode,
    this.data,
    this.message,
    this.stackTrace,
  });

  final bool ok;
  final int? statusCode;
  final dynamic data;
  final String? message;
  final StackTrace? stackTrace;

  factory ApiResponse.success(int statusCode, dynamic data) {
    return ApiResponse._(ok: true, statusCode: statusCode, data: data);
  }

  factory ApiResponse.error(String message, {int? statusCode, StackTrace? stackTrace, dynamic data}) {
    return ApiResponse._(
      ok: false,
      statusCode: statusCode,
      message: message,
      stackTrace: stackTrace,
      data: data,
    );
  }
}

File: src/services/apiClient.js
import axios from 'axios'

const BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000'
const TIMEOUT  = Number(import.meta.env.VITE_API_TIMEOUT) || 15000

let _accessToken  = sessionStorage.getItem('_at') || null
let _refreshToken = sessionStorage.getItem('_rt') || null
let _refreshing   = false

const http = axios.create({ baseURL: BASE_URL, timeout: TIMEOUT })

http.interceptors.request.use((config) => {
  if (_accessToken) config.headers['Authorization'] = `Bearer ${_accessToken}`
  return config
})

async function _tryRefresh() {
  if (!_refreshToken || _refreshing) return false
  _refreshing = true
  try {
    const res = await http.post('/api/auth/refresh', null, {
      headers: { Authorization: `Bearer ${_refreshToken}` },
    })
    const { access_token, refresh_token } = res.data
    if (access_token) {
      _accessToken = access_token
      sessionStorage.setItem('_at', _accessToken)
      if (refresh_token) {
        _refreshToken = refresh_token
        sessionStorage.setItem('_rt', _refreshToken)
      }
      return true
    }
  } catch (_) {}
  _accessToken  = null
  _refreshToken = null
  sessionStorage.removeItem('_at')
  sessionStorage.removeItem('_rt')
  return false
}

http.interceptors.response.use(
  (res) => res,
  async (err) => {
    const orig = err.config
    if (err.response?.status === 401 && _refreshToken && !orig._retry) {
      orig._retry = true
      _refreshing = false
      const ok = await _tryRefresh()
      _refreshing = false
      if (ok) {
        orig.headers['Authorization'] = `Bearer ${_accessToken}`
        return http(orig)
      }
    }
    _refreshing = false
    return Promise.reject(err)
  }
)

export function setAuthToken(token, refreshToken) {
  _accessToken = token || null
  if (refreshToken !== undefined) _refreshToken = refreshToken || null
  if (_accessToken) {
    sessionStorage.setItem('_at', _accessToken)
    if (_refreshToken) sessionStorage.setItem('_rt', _refreshToken)
    else sessionStorage.removeItem('_rt')
  } else {
    sessionStorage.removeItem('_at')
    sessionStorage.removeItem('_rt')
    sessionStorage.removeItem('_user')
  }
}

export function hasAuthToken() { return !!_accessToken }

export async function logout() {
  if (_accessToken) {
    try { await http.post('/api/auth/logout') } catch (_) {}
  }
  _accessToken  = null
  _refreshToken = null
  sessionStorage.removeItem('_at')
  sessionStorage.removeItem('_rt')
  sessionStorage.removeItem('_user')
}

export default http
